File: frontend/kafka-test.mjs
/**
 * Example: Kafka Producer Usage
 * This example demonstrates recommended patterns for using Kafka in Next.js.
 * 
 * For production usage, see:
 * - src/lib/kafka/client.ts - Recommended producer implementation
 * 
 * Usage:
 *   node kafka-test.mjs
 * 
 * Or in Next.js API route or component:
 *   import { getProducer } from '@/lib/kafka/client';
 */

import { KafkaProducerClient } from './src/lib/kafka/client.js';

async function run() {
  const producer = new KafkaProducerClient();
  
  try {
    // Connect to Kafka
    await producer.connect();
    console.log("Connected to Kafka");
    
    // Send a message using recommended pattern
    const metadata = await producer.sendMessage(
      'campaign.created.json',
      {
        campaign_id: '123',
        name: 'Example Campaign from Next.js',
        created_at: new Date().toISOString(),
      },
      '123', // Key for partitioning
      {
        source: 'nextjs-example',
        version: '1.0',
      }
    );
    
    console.log("Message sent successfully:", metadata);
    console.log(`Topic: ${metadata.topicName}, Partition: ${metadata.partition}, Offset: ${metadata.offset}`);
    
    // Example: Send batch messages
    const batchMetadata = await producer.sendBatch([
      {
        topic: 'campaign.created.json',
        value: { campaign_id: '124', name: 'Campaign 1' },
        key: '124',
      },
      {
        topic: 'campaign.created.json',
        value: { campaign_id: '125', name: 'Campaign 2' },
        key: '125',
      },
    ]);
    
    console.log(`Sent ${batchMetadata.length} messages in batch`);
    
  } catch (error) {
    console.error("Error:", error);
    process.exit(1);
  } finally {
    // Always disconnect
    await producer.disconnect();
    console.log("Disconnected from Kafka");
  }
}

run().catch(console.error);

File: backend/kafka_test.py
"""
Example: Kafka Producer and Consumer Usage
This example demonstrates recommended patterns for using Kafka in Django.

For production usage, see:
- backend/kafka_producer.py - Recommended producer implementation
- backend/kafka_consumer.py - Recommended consumer implementation
- docs/KAFKA_CLIENT_GUIDE.md - Complete usage guide
"""

from backend.kafka_producer import KafkaProducerClient
from backend.kafka_consumer import KafkaConsumerClient
import logging

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


def example_producer():
    """Example: Using the recommended Kafka producer pattern."""
    # Create producer using context manager (automatic cleanup)
    with KafkaProducerClient() as producer:
        # Send a message
        success = producer.send_message(
            topic='campaign.created.json',
            value={
                'campaign_id': '123',
                'name': 'Example Campaign',
                'created_at': '2024-01-15T10:00:00Z'
            },
            key='123',
            headers={
                'source': 'example-script',
                'version': '1.0'
            }
        )
        
        if success:
            logger.info("Message sent successfully")
        else:
            logger.error("Failed to send message")


def example_consumer():
    """Example: Using the recommended Kafka consumer pattern."""
    def handle_message(message):
        """Message handler - processes each message."""
        try:
            logger.info(f"Received message: {message}")
            
            # Extract data
            value = message['value']
            topic = message['topic']
            key = message['key']
            
            # Process message based on topic
            if 'campaign.created' in topic:
                logger.info(f"Processing campaign creation: {value}")
                # Your processing logic here
            
            return True  # Return True on success
        except Exception as e:
            logger.error(f"Error processing message: {e}")
            return False  # Return False on failure
    
    # Create consumer
    consumer = KafkaConsumerClient(
        topics=['campaign.created.json'],
        group_id='example-consumer-group',
        message_handler=handle_message
    )
    
    try:
        # Start consuming (this blocks until stopped)
        consumer.start(blocking=True)
    except KeyboardInterrupt:
        logger.info("Stopping consumer...")
    finally:
        consumer.stop()


if __name__ == '__main__':
    # Run producer example
    logger.info("Running producer example...")
    example_producer()
    
    # To run consumer example, uncomment:
    # logger.info("Running consumer example...")
    # example_consumer()
